benchmark/atoi-corpus.h

```diff
@@ -143,10 +145,171 @@ struct CorpusStringLength {
     AVX2_STRLEN_CORPUS_X_LIST \
     NEON_STRLEN_CORPUS_X_LIST
 
+struct CorpusLeadingSpaces {
+    constexpr static auto CountOfSpaceCharactersAvailable = 6;
+    constexpr static inline std::array<char, CountOfSpaceCharactersAvailable> Spaces =
+        { ' ', '\n', '\t', '\r', '\f', '\v' };
+    std::vector<int> skips_;
+    std::string characters_;
+
+    CorpusLeadingSpaces(std::vector<int> &&skips, std::string &&cs):
+        skips_{std::move(skips)}, characters_{std::move(cs)}
+    {}
+
+    template<typename G>
+    static auto makeCorpus(G &generator) {
+        auto count = 1031; // see Corpus8DecimalDigits for why 1031
+        std::vector<int> sizes;
+        std::string allCharacters;
+        std::geometric_distribution<>
+            spacesCount(1.0/29),
+            extraCharacters(0.5);
+            // unrepresentatively very large, but will cross the 32 boundary
+            // to test 32-byte techniques
+        std::uniform_int_distribution<>
+            spacer(0, CountOfSpaceCharactersAvailable - 1),
+            moreCharacters(0, 255);
+
+        while(count--) {
+            auto count = spacesCount(generator);
+            for(auto i = count; i--; ) {
+                allCharacters.append(1, Spaces[spacer(generator)]);
+            }
+            auto extra = moreCharacters(generator);
+            for(auto i = extra; i--; ) {
+                allCharacters.append(1, moreCharacters(generator));
+            }
+            sizes.push_back(count + extra + 1);
+            allCharacters.append(1, '\0');
+        }
+        return CorpusStringLength(std::move(sizes), std::move(allCharacters));
+    }
+
+    struct Iterator {
+        int *skips, *sentinel;
+        char *cp;
+
+        Iterator &operator++() {
+            cp += *skips++;
+            return *this;
+        }
+
+        char *operator*() {
+            return cp;
+        }
+
+        auto next() noexcept {
+            ++(*this);
+            return sentinel != skips;
+        }
+    };
+
+    Iterator commence() {
```

Comment:
why commence instead of begin or start? Seems like an unusual name, so probably warrants an explanatory comment.

Reply:
I hope you strongly agree with all of these reasons:

1. I'd like everybody to be able to express their ideas in the way most natural to them, without being encumbered with justifications.  In this way, we maximize the benefit of a plurality of ways to express things.  For example, you introduced "Lanes" as part of the nomenclature of SWAR, I was fastidiously talking of "sub elements", trying to avoid the confusion of people thinking the library used true support for lanes in the way of assembler specific instructions, or something like that, until "sub elements" began to feel insufferable, I adopted "Lanes" and I'm "running" with it.  Of course I was scared of something I should not have been scared by thinking people might get confused here.  Obviously, the whole point of the SWAR is to have "virtual" lanes, nobody will misunderstand this.
1.1. It is important to not require justifications.  Otherwise the default is "always talk in rhyme" as in Tony Van Eerd's presentation, style is not for the code to have a regularity without practical benefit, style is to facilitate the programmer to express ideas, hence it is subjective.  If you choose to indent to two spaces while most of the code is indented at four spaces, I will explain to you the reason against that it introduces accidental levels of indentation, something I avoid not just with the number of spaces, but semantically (I don't indent to the start of the arguments in a function call), the idea is that, in the absence of a superior recommendation, I find that making the indentation level being the same as the semantical nesting level (not syntactical) is helpful to comprehension.
2. In the practical sense, I don't think it helps productivity to explain everything that is unusual.  Otherwise, we'd have to start by having an idea of how unusual things must be before having to be justified.
3. By explaining trivial things explicitly, one devalues the value of important explanations.  A comment is an imposition on everyone: "I write this because I think it is necessary, thus, if you see my comments, read them", then you are forced to notice if the comment changed or if the code invalidated the comment during its evolution.
4. `begin` or `start` might evoke the wrong style of iteration, this is Java's.  I think `commence` is not used elsewhere, then it is perfect to denote "an iteration style that is not the usual style in C++".
5. `Iterator` is used because that's what it is, however, the use of `Iterator` is not important in this code because this type is supposed to be used as an opaque type by the user, all the knowledge needed is to see that termination is not determined, from the programmer's perspective, by the comparison against a sentinel, but the falseness of `next()`.

Could you articulate better why an unusual name may warrant an explanation?

Reply:
"""begin or start might evoke the wrong style of iteration, this is Java's. I think commence is not used elsewhere, then it is perfect to denote "an iteration style that is not the usual style in C++".""" -> 'we use commence to denote an iteration style that isn't the usual C++ iteration style'.

Your points are valid, but a new to c++ or not c++ focused reader of this would certainly benefit from a short comment explicitly saying "This isn't the usualy C++ iterator!".  My request was only to benefit them.

Based on size of response, maybe you don't want to do it. That's ok too.

Reply:
I still believe in point 3 above, it is counterproductive to explain the easy things.
The length of the response is to make clear that we want to allow people to express themselves and deal with the inconsistencies as readers and reviewers and maintainers.
I think the discussion is important, you mention two things: unusual things may warrant an explanation, we agree a short comment might benefit the new to C++ or not C++ focused reader, but how much do we have to cater to them?
My use of `commence` caught your attention, it did its job.
A common saying from Spanish: "A buen entendedor, pocas palabras", "to a good understander, few words".